### DISCUSSION TITLE
Clarification of custom runners

### DISCUSSION THREAD

**Discussion by sadilekivan** [cross-rs/cross#1173] [Q&A] [answered]
My intentions are to cross compile for a remote RaspberryPi board and with cargo I can simply specify a custom cargo runner that uploads the file and runs it over ssh. This even gets the output back on terminal.

I'm trying to do the same with a cross runner but I cannot specify a custom file/script/binary. I get an error:
```
Invalid runner: "~/.config/nushell/scripts/cargo-remote-runner"
Valid runners are: native, qemu-user and qemu-system
```
I could not find more info about these runners and what are their differences. Furthermore how would I go about calling a custom runner as specified in the [`README.md`](https://github.com/cross-rs/cross#option-1-configuring-cross-directly-in-your-cargotoml)

**Comment by Emilgardis** [comment, marked as answer]
Runners work the same for cross as with cargo, however you'd need to mount the executable so that it's visible in the docker container. So, instead of using the `runner` option, use `CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER` directly.

The only things that are mounted is the current directory, and the workspace directory, (e.g the folder with the lowest `Cargo.toml`)

You can mount arbitrary folders with https://github.com/cross-rs/cross/wiki/Configuration#targettargetenv

I do notice that you're using nushell as well, and that would mean that you'd have to install nushell in the docker image too or make the script work in bash (note that cargo is not included in the image at the [`pre-build` step](https://github.com/cross-rs/cross/wiki/Configuration#targettargetpre-build)) . Full solution (with a fixed bash version of `cargo-remote-runner`) would be something like

```toml
[target.aarch64-unknown-linux-gnu.env]
passthrough = ["CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER=./cargo-remote-runner"]
```

**Reply by Emilgardis** [reply]
The `runner` option in `Cross.toml` only take predetermined runners, like qemu. Predetermined runners are defined [here](https://github.com/cross-rs/cross/blob/f0190b0c0fd8fc354d9b0fad61c06bf0f29be277/targets.toml#L55)

**Reply by sadilekivan** [reply]
If anyone happens to be attempting the same, here is my solution. I haven't gotten `.local` hostnames to work but IPs do so far. I've added `rsync` to the container and am using following.

`Cross.toml`
```
[target.aarch64-unknown-linux-gnu.env]
passthrough = [
    "CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER=./remote-runner",
    "CARGO_TARGET_HOSTNAME=root@10.0.0.1"
]
```

`remote-runner`
```
#!/bin/bash
local_path=$1
name=$(basename $local_path)
remote_path="/tmp/$name"
ssh_flags="-o \"UserKnownHostsFile=/dev/null\" -o \"StrictHostKeyChecking no\" -o \"LogLevel ERROR\""

transfer_command="rsync $local_path $CARGO_TARGET_HOSTNAME:$remote_path -e 'ssh $ssh_flags'"
execute_command="ssh $ssh_flags -t $CARGO_TARGET_HOSTNAME $remote_path"

eval $transfer_command && \
eval $execute_command
```
